### PR TITLE
doc: Add some better examples for scripted diff

### PR DESCRIPTION
**Pull request by laanwj**
The current example isn't too great, for example it uses `find` instead of `git ls-files`. Add a subsection with suggestions and examples.

Feel free to propose some other great examples to add.

### CODE REVIEW

File: doc/developer-notes.md [hunk 1]
```diff
@@ -910,7 +910,35 @@ For development, it might be more convenient to verify all scripted-diffs in a r
...
+- [scripted-diff: Remove g_connman, g_banman globals](https://github.com/bitcoin/bitcoin/commit/301bd41a2e6765b185bd55f4c541f9e27aeea29d)
+replaces specific terms in a list of specific source files.
+
+- [scripted-diff: Replace fprintf with tfm::format](https://github.com/bitcoin/bitcoin/commit/fac03ec43a15ad547161e37e53ea82482cc508f9)
+does a global replacement but excludes certain directories.
+
+To find all previous uses of scripted diffs in the repository, do:
+
+```
+git log --grep="-BEGIN VERIFY SCRIPT-"
```
Comment (promag) [inline doc/developer-notes.md]: Great tip!

**Review by promag — COMMENTED** [review]
Concept ACK.

You have to update the TOC.

**Comment by practicalswift** [comment]
Concept ACK -- `git ls-files` and `git grep` are both great! :)

**Review by maflcko — APPROVED** [review]

**Comment by laanwj** [comment]
BTW: what is the best way to do a re-apply of a conflicted scripted diff when rebasing? I did the following to rebase #17410 and it seems it could be more efficient, and if so, we'd want to add it to the documentation:

```bash
$ git rebase master
First, rewinding head to replay your work on top of it...                                      
Applying: Rename `db` log category to `walletdb` (like `coindb`)                                          
Applying: scripted-diff: Change `BCLog::DB` to `BCLog::WALLETDB`                                          Using index info to reconstruct a base tree...                                                            
M       src/wallet/db.cpp                                                                          
Falling back to patching base and 3-way merge...                                        
Auto-merging src/wallet/db.cpp                                                                 
CONFLICT (content): Merge conflict in src/wallet/db.cpp                         
error: Failed to merge in the changes.                                                                    
Patch failed at 0002 scripted-diff: Change `BCLog::DB` to `BCLog::WALLETDB`               
Use 'git am --show-current-patch' to see the failed patch               
                                                                                                          
Resolve all conflicts manually, mark them as resolved with                                 
"git add/rm <conflicted_files>", then run "git rebase --continue".                             
You can instead skip this commit: run "git rebase --skip".                                                
To abort and get back to the state before "git rebase", run "git rebase --abort". 
```
do hard reset, to get rid of git's attempt at re-applying, and conflict markers
```
$ git reset --hard
```
manually paste script steps
```
$ git grep -l "BCLog::DB" src | xargs sed -i "s/BCLog::DB/BCLog::WA
LLETDB/g"
$ sed -i "s/DB          =/WALLETDB    =/g" src/logging.h
$ git status
```
mandually  git add changed objects, and roll rebase forward
```
$ git add src/logging.cpp src/logging.h src/wallet/db.cpp src/walle
t/walletdb.cpp
$ git rebase --continue
```

**Comment by ryanofsky** [comment]
I have a one-line [`script-get`](https://github.com/ryanofsky/home/blob/5268deca53b259d42ac70980ece7c1c44b10fd8a/src/pr.sh#L622-L624) bash function would let you replace the manual paste step:

```sh
bash -c "$(script-get HEAD)"
```

Probably more of this could be scripted in general, though

**Comment by laanwj** [comment]
Thanks!

Indeed, my idea was that this is pretty much an alternative merge strategy: ignore git's own attempt (which could cause conflicts or miss new occurences) in favor of re-running the script.
It won't have any merge conflicts, by definition, though obviously the result still has to be reviewed.

File: doc/developer-notes.md [hunk 2]
```diff
@@ -910,7 +910,35 @@ For development, it might be more convenient to verify all scripted-diffs in a r
 test/lint/commit-script-check.sh origin/master..HEAD
 ```
 
-Commit [`bb81e173`](https://github.com/bitcoin/bitcoin/commit/bb81e173) is an example of a scripted-diff.
+### Suggestions and examples
```
Comment (hebasto) [inline doc/developer-notes.md]: nit: TOC could be updated as @promag [suggests](https://github.com/bitcoin/bitcoin/pull/17411#pullrequestreview-314052877).

File: doc/developer-notes.md [hunk 3]
```diff
@@ -910,7 +910,35 @@ For development, it might be more convenient to verify all scripted-diffs in a r
...
+
+If you need to replace in multiple files, prefer `git ls-files` to `find` or globbing, and `git grep` to `grep`, to
+avoid changing files that are not under version control.
+
+For efficient replacement scripts, reduce the selection to the files that potentially need to be modified, so for
+example, instead of a blanket `git ls-files src | xargs sed -i s/apple/orange/`, use
+`git grep -l apple src | xargs sed -i s/apple/orange/`.
+
+Also, it's good to keep the selection of files as specific as possible — for example, replace only in directories where
```
Comment (hebasto) [inline doc/developer-notes.md]: nit: contractions, "it's", should be avoided in well-styled docs ;)

**Review by hebasto — APPROVED** [review]
ACK d8ae24fae2fce7f39a9eb7fa7c9b21fac79e82ac, I have reviewed the code and it looks OK, I agree it can be merged.

File: doc/developer-notes.md (hunk 2 above)
Comment (laanwj) [reply doc/developer-notes.md]: yes, will update the TOC

File: doc/developer-notes.md (hunk 3 above)
Comment (laanwj) [reply doc/developer-notes.md]: ok

**Comment by laanwj** [comment]
Updated toc, replaced occurrences of "it's" with "it is".

**Review by hebasto — APPROVED** [review]
re-ACK adbe15504713ddba6e9c024c59d977675d49e350

**Pull request merged**